File: apps/monitoring-sidecar/tests/unit/audio-meter-crosscheck.test.ts
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { describe, expect } from 'vitest';

import { decodeWav } from '@scribear/test-audio-source';

/**
 * PLAN-AUDIOVIZ §9 cross-check gate, standalone-page leg.
 *
 * `audio-meter-dsp.test.ts` already gates this page's maths against tones it
 * defines itself. This suite is different: it reads the *shared* expectation
 * table in `tools/audio-meter-crosscheck/fixtures.json`, the same table the
 * Python publisher's meter is held to, so that the two surfaces an operator
 * compares side by side are pinned to one set of numbers rather than to two
 * independently-written ones.
 *
 * See `tools/audio-meter-crosscheck/README.md` for what this does and does not
 * cover — notably, it stops at the DSP and does not exercise the live transport.
 */

const REPO_ROOT = new URL('../../../../', import.meta.url);

const HTML_PATH = fileURLToPath(
  new URL('libs/audio-meter-page/audio-meter.html', REPO_ROOT),
);
const FIXTURES_PATH = fileURLToPath(
  new URL('tools/audio-meter-crosscheck/fixtures.json', REPO_ROOT),
);

interface Expected {
  rmsDbfs: number;
  peakDbfs: number;
  clippingPct: number;
}

interface ToneFixture {
  name: string;
  rmsDbfs: number;
  frequencyHz: number;
  sampleRate: number;
  seconds: number;
  expected: Expected;
}

interface WavFixture {
  path: string;
  sampleRate: number;
  sampleCount: number;
  expected: Expected;
}

interface LimitedToneFixture extends ToneFixture {
  /** Gain applied before hard-limiting. */
  preLimitGain: number;
  /** Ceiling the signal is limited to; defaults to full scale. */
  limitCeiling?: number;
}

interface Fixtures {
  toleranceDb: number;
  wav: WavFixture;
  tones: ToneFixture[];
  limitedTones: LimitedToneFixture[];
}

interface Readings {
  rmsDb: number;
  heldPeakDb: number;
  peakDb: number;
  maxTruePeakDb: number;
  clippingPercent: number;
}

interface MeterCore {
  process(block: Float32Array): void;
  read(): Readings;
}

interface Dsp {
  AudioMeterCore: new (
    sampleRate: number,
    options?: Record<string, number>,
  ) => MeterCore;
}

/** Same extraction as `audio-meter-dsp.test.ts`: evaluate the page's own DSP
 *  block so the shipped text and the tested maths cannot drift apart. */
function loadDsp(): Dsp {
  const html = readFileSync(HTML_PATH, 'utf8');
  const match = /<script id="meter-dsp">([\s\S]*?)<\/script>/.exec(html);
  if (!match?.[1])
    throw new Error('meter-dsp script block not found in the page');
  const sandbox: { ScribeArAudioMeter?: Dsp } = {};
  // eslint-disable-next-line @typescript-eslint/no-implied-eval
  const install = new Function('globalThis', match[1]) as (
    scope: typeof sandbox,
  ) => void;
  install(sandbox);
  if (!sandbox.ScribeArAudioMeter)
    throw new Error('DSP block exported nothing');
  return sandbox.ScribeArAudioMeter;
}

/**
 * Decodes the fixture WAV with the sidecar's own reader — the same
 * `decodeWav` the synthetic canary uses to put fixture audio on the wire, so
 * this leg reads the file exactly as production does rather than through a
 * second parser written for the test.
 */
function readFixtureWav(path: string): {
  sampleRate: number;
  samples: Float32Array;
} {
  const decoded = decodeWav(readFileSync(path));
  if (decoded.channels !== 1) {
    throw new Error(`expected mono, got ${String(decoded.channels)} channels`);
  }
  const samples = new Float32Array(Math.floor(decoded.pcm.length / 2));
  for (let i = 0; i < samples.length; i += 1) {
    samples[i] = decoded.pcm.readInt16LE(i * 2) / 32768;
  }
  return { sampleRate: decoded.sampleRate, samples };
}

const BLOCK = 128;

/** Feeds samples through the core in worklet-sized blocks, as the page does. */
function feed(core: MeterCore, samples: Float32Array): void {
  for (let written = 0; written < samples.length; written += BLOCK) {
    core.process(
      samples.subarray(written, Math.min(written + BLOCK, samples.length)),
    );
  }
}

function sineSamples(fixture: ToneFixture): Float32Array {
  // A = 10^(dBFS/20) * sqrt(2), so the sine's RMS (A/sqrt(2)) is the target.
  const amplitude = Math.pow(10, fixture.rmsDbfs / 20) * Math.SQRT2;
  const total = Math.round(fixture.seconds * fixture.sampleRate);
  const out = new Float32Array(total);
  for (let i = 0; i < total; i += 1) {
    out[i] =
      amplitude *
      Math.sin((2 * Math.PI * fixture.frequencyHz * i) / fixture.sampleRate);
  }
  return out;
}

const dsp = loadDsp();
const fixtures = JSON.parse(readFileSync(FIXTURES_PATH, 'utf8')) as Fixtures;
const TOLERANCE_DB = fixtures.toleranceDb;

describe('cross-check: shared tone fixtures', (it) => {
  for (const tone of fixtures.tones) {
    it(`reads "${tone.name}" within the shared tolerance`, () => {
      // Arrange
      const core = new dsp.AudioMeterCore(tone.sampleRate);

      // Act
      feed(core, sineSamples(tone));
      const readings = core.read();

      // Assert — the same numbers audio_meter.py is held to for this fixture.
      expect(Math.abs(readings.rmsDb - tone.expected.rmsDbfs)).toBeLessThan(
        TOLERANCE_DB,
      );
      expect(
        Math.abs(readings.heldPeakDb - tone.expected.peakDbfs),
      ).toBeLessThan(TOLERANCE_DB);
      expect(readings.clippingPercent).toBe(tone.expected.clippingPct * 100);
    });
  }
});

/** Feeds one full metering window of the fixture WAV and reads the meter. */
function readWavFixture(): Readings {
  const { path, sampleRate, sampleCount } = fixtures.wav;
  const wav = readFixtureWav(fileURLToPath(new URL(path, REPO_ROOT)));
  expect(wav.sampleRate).toBe(sampleRate);
  expect(wav.samples.length).toBeGreaterThanOrEqual(sampleCount);

  const core = new dsp.AudioMeterCore(sampleRate);
  feed(core, wav.samples.subarray(0, sampleCount));
  return core.read();
}

describe('cross-check: known speech WAV', (it) => {
  it('reads the excerpt RMS within the shared tolerance', () => {
    // Arrange / Act — exactly one full metering window of real speech, so the
    // reading averages precisely the samples the expectation was computed over.
    const readings = readWavFixture();

    // Assert — an arithmetic expectation, not a value copied from either meter.
    // RMS is the number the dashboard's meter bar renders, so this is the
    // assertion that makes the two surfaces agree on what the operator sees.
    expect(
      Math.abs(readings.rmsDb - fixtures.wav.expected.rmsDbfs),
    ).toBeLessThan(TOLERANCE_DB);
  });

  it('reads the excerpt peak within the shared tolerance', () => {
    // Act
    const readings = readWavFixture();

    // Assert — `maxTruePeakDb` is the field comparable to the publisher's
    // `peak_dbfs` (`max|x|` over the window): both are a maximum over
    // everything seen. It is an oversampled reconstruction rather than a sample
    // maximum, so it does not bracket the sample peak from either side — the
    // polyphase filter's passband loss puts it ~0.001 dB *below* it on this
    // excerpt, well inside the gate.
    expect(
      Math.abs(readings.maxTruePeakDb - fixtures.wav.expected.peakDbfs),
    ).toBeLessThan(TOLERANCE_DB);
  });

  it('exposes no clipping for an excerpt that has none', () => {
    expect(readWavFixture().clippingPercent).toBe(
      fixtures.wav.expected.clippingPct * 100,
    );
  });

  it('does not expose a window-max sample peak — `peakDb` is a recent-peak meter', () => {
    // A divergence worth pinning down rather than papering over: the publisher's
    // `peak_dbfs` is the max over its whole window, but this page has no such
    // field. `peakDb` / `heldPeakDb` are a short peak window with hold-and-decay
    // (a broadcast-style peak meter), so on a long excerpt they report the
    // *recent* peak and sit well below the excerpt maximum. Anyone comparing the
    // two surfaces' "Peak" readings is comparing different measurements; this
    // asserts the difference is real so nobody "fixes" the cross-check by
    // pointing it at the wrong field.
    const readings = readWavFixture();

    expect(readings.peakDb).toBeLessThan(
      fixtures.wav.expected.peakDbfs - TOLERANCE_DB,
    );
    expect(readings.heldPeakDb).toBeLessThan(
      fixtures.wav.expected.peakDbfs - TOLERANCE_DB,
    );
  });
});

/** Hard-limits a tone to +/-`limitCeiling` after applying its pre-limit gain.
 *  A ceiling below CLIP_THRESHOLD yields a flat plateau that must NOT be charged. */
function limitedSineSamples(fixture: LimitedToneFixture): Float32Array {
  const ceiling = fixture.limitCeiling ?? 1;
  // `map` rather than an indexed write loop: TypeScript 6 types indexed reads on
  // a typed array as `number | undefined`, and a `?? 0` fallback here would
  // silently substitute silence for a bug rather than failing.
  return sineSamples(fixture).map((sample) =>
    Math.max(-ceiling, Math.min(ceiling, fixture.preLimitGain * sample)),
  );
}

describe('cross-check: hard-limited audio', (it) => {
  /*
   * The counterpart to the clean full-scale sine above. Requiring a run at the
   * rail is only the right rule if it still catches real clipping, so these pin
   * the positive cases too — and one fixture whose plateau sits *below* the
   * threshold, which pins the threshold from below. Between them, both halves of
   * the shared rule (threshold and minimum run) fail a test if either drifts.
   */
  for (const tone of fixtures.limitedTones) {
    it(`charges "${tone.name}" the shared fraction`, () => {
      // Arrange
      const core = new dsp.AudioMeterCore(tone.sampleRate);

      // Act
      feed(core, limitedSineSamples(tone));
      const readings = core.read();

      // Assert — an exact comparison: clippingPct is a counted fraction of
      // samples, not a measurement, so the two implementations must agree to
      // the digit rather than within a dB tolerance.
      expect(readings.clippingPercent).toBeCloseTo(
        tone.expected.clippingPct * 100,
        4,
      );
      expect(Math.abs(readings.rmsDb - tone.expected.rmsDbfs)).toBeLessThan(
        TOLERANCE_DB,
      );
    });
  }

  it('puts genuinely clipped audio past the dashboard crit threshold', () => {
    // The rule must still be actionable: a red "clipping" chip fires at 1%.
    const clipping = fixtures.limitedTones.filter(
      (t) => t.expected.clippingPct > 0,
    );

    expect(clipping.length).toBeGreaterThan(0);
    for (const tone of clipping) {
      expect(tone.expected.clippingPct).toBeGreaterThan(0.01);
    }
  });
});

File: transcription_service/tests/unit/shared/utils/audio_meter/audio_meter_crosscheck_test.py
"""
PLAN-AUDIOVIZ §9 cross-check gate, publisher leg

`audio_meter_test.py` already gates this meter against tones it synthesizes
itself. This suite is different: it reads the *shared* expectation table in
`tools/audio-meter-crosscheck/fixtures.json`, the same table the standalone
meter page's DSP is held to, so the two surfaces an operator compares side by
side are pinned to one set of numbers rather than to two independently-written
ones.

Neither meter is the oracle. Every expectation is arithmetic - the RMS and peak
of the exact sample sequence each fixture defines - so a failure here means this
meter is wrong, not that it disagrees with its counterpart's opinion.

See `tools/audio-meter-crosscheck/README.md` for what this gate does and does
not cover; notably it stops at the DSP and does not exercise the live transport.
"""

import json
import wave
from pathlib import Path

import numpy as np
import pytest

from src.shared.utils.audio_meter import AudioLevelStats, AudioMeter


def _repo_root() -> Path:
    """
    Walks up to the directory holding both the fixture manifest and the WAV.

    The suite's working directory depends on where pytest was invoked from, so
    neither it nor a fixed number of `parent` hops is assumed.
    """
    for candidate in [
        Path(__file__).resolve(),
        *Path(__file__).resolve().parents,
    ]:
        if (candidate / "tools/audio-meter-crosscheck/fixtures.json").is_file():
            return candidate
    raise AssertionError(
        "Could not locate tools/audio-meter-crosscheck/fixtures.json above "
        f"{__file__}"
    )


REPO_ROOT = _repo_root()
FIXTURES = json.loads(
    (REPO_ROOT / "tools/audio-meter-crosscheck/fixtures.json").read_text(
        encoding="utf-8"
    )
)
TOLERANCE_DB = FIXTURES["toleranceDb"]


def _sine(fixture: dict) -> np.ndarray:
    """
    Generates a fixture's tone.

    A = 10^(dBFS/20) * sqrt(2), so the sine's RMS (A/sqrt(2)) lands on the
    requested level. Identical arithmetic to the JavaScript leg's `sineSamples`,
    so both meters see the same sample sequence.
    """
    total = round(fixture["seconds"] * fixture["sampleRate"])
    amplitude = 10 ** (fixture["rmsDbfs"] / 20) * np.sqrt(2)
    index = np.arange(total)
    return (
        amplitude
        * np.sin(
            2 * np.pi * fixture["frequencyHz"] * index / fixture["sampleRate"]
        )
    ).astype(np.float32)


def _wav_excerpt() -> tuple[int, np.ndarray]:
    """The fixture WAV's leading excerpt, as float32 in [-1, 1)."""
    spec = FIXTURES["wav"]
    with wave.open(str(REPO_ROOT / spec["path"])) as handle:
        assert handle.getnchannels() == 1, "fixture WAV must be mono"
        assert handle.getsampwidth() == 2, "fixture WAV must be 16-bit PCM"
        assert handle.getframerate() == spec["sampleRate"]
        raw = handle.readframes(handle.getnframes())

    samples = np.frombuffer(raw, dtype="<i2").astype(np.float32) / 32768.0
    assert len(samples) >= spec["sampleCount"]
    return spec["sampleRate"], samples[: spec["sampleCount"]]


def _snapshot(sample_rate: int, samples: np.ndarray) -> AudioLevelStats:
    """Feeds one full metering window and reads the meter."""
    seconds = len(samples) / sample_rate
    meter = AudioMeter(sample_rate=sample_rate, window_sec=seconds)
    meter.append(samples)
    snapshot = meter.snapshot()
    assert snapshot is not None
    return snapshot


class TestSharedToneFixtures:
    """Gate: the shared tone table, read within the shared tolerance."""

    @pytest.mark.parametrize(
        "fixture", FIXTURES["tones"], ids=[t["name"] for t in FIXTURES["tones"]]
    )
    def test_reads_tone_within_shared_tolerance(self, fixture):
        """Each shared tone's rms/peak dBFS agree with the arithmetic value."""
        # Arrange
        samples = _sine(fixture)

        # Act
        snapshot = _snapshot(fixture["sampleRate"], samples)

        # Assert - the same numbers the standalone page's DSP is held to.
        expected = fixture["expected"]
        assert abs(snapshot.rms_dbfs - expected["rmsDbfs"]) < TOLERANCE_DB
        assert abs(snapshot.peak_dbfs - expected["peakDbfs"]) < TOLERANCE_DB
        # clippingPct is absent for fixtures the two implementations disagree
        # about; those are pinned per-side in TestKnownDivergences instead.
        if "clippingPct" in expected:
            assert snapshot.clipping_pct == pytest.approx(
                expected["clippingPct"]
            )


class TestSharedWavFixture:
    """Gate: a known speech WAV, read within the shared tolerance."""

    def test_reads_excerpt_rms_within_shared_tolerance(self):
        """
        The excerpt's rms_dbfs matches its arithmetic RMS.

        This is the number the dashboard's meter bar renders, so it is the
        assertion that makes the two surfaces agree about what the operator
        sees.
        """
        # Arrange
        sample_rate, samples = _wav_excerpt()

        # Act
        snapshot = _snapshot(sample_rate, samples)

        # Assert
        expected = FIXTURES["wav"]["expected"]
        assert abs(snapshot.rms_dbfs - expected["rmsDbfs"]) < TOLERANCE_DB

    def test_reads_excerpt_peak_within_shared_tolerance(self):
        """The excerpt's peak_dbfs matches max|x| over the same samples."""
        # Arrange
        sample_rate, samples = _wav_excerpt()

        # Act
        snapshot = _snapshot(sample_rate, samples)

        # Assert - peak_dbfs is a window maximum. The page's comparable field is
        # `maxTruePeakDb`, not its hold-and-decay `peakDb`; see the JS leg.
        expected = FIXTURES["wav"]["expected"]
        assert abs(snapshot.peak_dbfs - expected["peakDbfs"]) < TOLERANCE_DB

    def test_reports_no_clipping_for_an_excerpt_that_has_none(self):
        """Real speech at -26 dBFS RMS charges nothing as clipping."""
        # Arrange
        sample_rate, samples = _wav_excerpt()

        # Act
        snapshot = _snapshot(sample_rate, samples)

        # Assert
        expected = FIXTURES["wav"]["expected"]
        assert snapshot.clipping_pct == pytest.approx(expected["clippingPct"])

    def test_real_speech_is_not_silence(self):
        """
        The excerpt must not read as silence.

        A cross-check that only compared levels could pass while the silence
        flag was inverted - and `silence` is what drives the dashboard's crit
        chip, so a wrong value here is a fleet-wide false alarm.
        """
        # Arrange
        sample_rate, samples = _wav_excerpt()

        # Act
        snapshot = _snapshot(sample_rate, samples)

        # Assert
        assert snapshot.silence is False


def _limited_sine(fixture: dict) -> np.ndarray:
    """
    The fixture's tone, hard-limited after its pre-limit gain

    `limitCeiling` defaults to full scale. A ceiling below 1.0 puts the plateau
    inside the CLIP_THRESHOLD..1.0 band, which is what pins that constant.
    """
    ceiling = fixture.get("limitCeiling", 1.0)
    return np.clip(
        fixture["preLimitGain"] * _sine(fixture).astype(np.float64),
        -ceiling,
        ceiling,
    ).astype(np.float32)


class TestSharedLimitedFixtures:
    """
    Gate: hard-limited audio is charged identically on both sides.

    The counterpart to the clean full-scale sine above. Requiring a run at the
    rail is only the right rule if it still catches real clipping, so these pin
    the positive cases too - and one fixture whose plateau sits *below* the
    threshold, which pins the threshold from below. Between them, both halves of
    the shared rule (threshold and minimum run) fail a test if either drifts.
    """

    @pytest.mark.parametrize(
        "fixture",
        FIXTURES["limitedTones"],
        ids=[t["name"] for t in FIXTURES["limitedTones"]],
    )
    def test_charges_the_shared_fraction(self, fixture):
        """A hard-limited tone's clipping_pct matches the shared value."""
        # Arrange
        samples = _limited_sine(fixture)

        # Act
        snapshot = _snapshot(fixture["sampleRate"], samples)

        # Assert - exact, not within a tolerance: clipping_pct is a counted
        # fraction of samples, so the two implementations must agree to the
        # digit rather than within a dB.
        expected = fixture["expected"]
        assert snapshot.clipping_pct == pytest.approx(
            expected["clippingPct"], abs=1e-6
        )
        assert abs(snapshot.rms_dbfs - expected["rmsDbfs"]) < TOLERANCE_DB

    def test_clipping_fixtures_clear_the_dashboard_crit_threshold(self):
        """
        The rule must stay actionable, not merely quiet.

        A red "clipping" chip fires at 1%, so every fixture that is supposed to
        represent real clipping has to land above it.
        """
        # Arrange
        clipping = [
            f
            for f in FIXTURES["limitedTones"]
            if f["expected"]["clippingPct"] > 0
        ]

        # Assert
        assert clipping
        for fixture in clipping:
            assert fixture["expected"]["clippingPct"] > 0.01
